### PR TITLE
Delete custom resources when apibinding is deleted

### PR DESCRIPTION
**Pull request by qiujian16**
Signed-off-by: Jian Qiu <jqiu@redhat.com>

## Summary

add controller to delete all CRs with apibinding is deleting

followup: avoid CR creation when apibinding is deleting

## Related issue(s)

Fixes https://github.com/kcp-dev/kcp/issues/846

### CODE REVIEW

File: pkg/reconciler/apis/apibindingdeletion/apibinding_deletor.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+						continue
+					}
+					finalizersToNumRemaining[finalizer] = finalizersToNumRemaining[finalizer] + numRemaining
+				}
+			}
+		}
+	}
+
+	if len(deleteContentErrs) > 0 {
+		conditions.MarkFalse(
```
Comment (stevekuznetsov) [inline pkg/reconciler/apis/apibindingdeletion/apibinding_deletor.go:95]: This side-effect mutation makes the logical flow hard to follow. Can we make this more explicit?

File: pkg/reconciler/apis/apibindingdeletion/apibinding_deletor.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+	deleteErrors := []error{}
+
+	for _, item := range partialList.Items {
+		if deletedNamespaces.Has(item.GetNamespace()) {
+			continue
+		}
+
+		// don't retry deleting the same namespace
+		deletedNamespaces.Insert(item.GetNamespace())
+		background := metav1.DeletePropagationBackground
```
Comment (stevekuznetsov) [inline pkg/reconciler/apis/apibindingdeletion/apibinding_deletor.go:170]: Why do we want this and the estimation logic for synthetic re-queues, instead of foreground? Unused goroutines as queue workers to handle burst are cheap, right?

File: pkg/reconciler/apis/apibindingdeletion/apibinding_deletor.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+			continue
+		}
+	}
+
+	deleteError := utilerrors.NewAggregate(deleteErrors)
+	if deleteError != nil {
+		return gvrDeletionMetadata{}, err
+	}
+
+	// list again so we know how many left resources
```
Comment (stevekuznetsov) [inline pkg/reconciler/apis/apibindingdeletion/apibinding_deletor.go:186]: This seems brittle, we just hope that during the time taken to issue the background delete calls some have finished?

File: test/e2e/apibinding/apibinding_deletion_test.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+	t.Logf("apibinding should have BindingResourceDeleteSuccess with false status")
+	require.Eventually(t, func() bool {
+		apibinding, err := kcpClients.Cluster(consumerWorkspace).ApisV1alpha1().APIBindings().Get(ctx, apiBinding.Name, metav1.GetOptions{})
+		if err != nil {
+			return false
+		}
+
+		return conditions.IsFalse(apibinding, apisv1alpha1.BindingResourceDeleteSuccess)
+	}, wait.ForeverTestTimeout, 1*time.Second)
```
Comment (stevekuznetsov) [inline test/e2e/apibinding/apibinding_deletion_test.go:175]: 1s seems like a long poll interval, why?

File: pkg/reconciler/apis/apibindingdeletion/apibinding_deletion_controller.go [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+	}
+	if len(apibinding.Finalizers) == len(copiedFinalizers) {
+		return nil
+	}
+
+	finalizerBytes, err := json.Marshal(copiedFinalizers)
+	if err != nil {
+		return err
+	}
+	patch := fmt.Sprintf("{\"metadata\": {\"finalizers\": %s}}", string(finalizerBytes))
```
Comment (stevekuznetsov) [inline pkg/reconciler/apis/apibindingdeletion/apibinding_deletion_controller.go:354]: OK, then perhaps two cases:

1. finalizers is empty -> use a static, back-ticks string
2. not empty -> use JSON Patch

File: pkg/reconciler/apis/apibindingdeletion/apibinding_deletion_controller.go [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+}
+
+func (c *Controller) patchFinalizer(ctx context.Context, cluster logicalcluster.Name, name string, finalizers []string) error {
+	finalizerData := &metav1.PartialObjectMetadata{
+		ObjectMeta: metav1.ObjectMeta{
+			Finalizers: finalizers,
+		},
+	}
+
+	patch, err := json.Marshal(finalizerData)
```
[the remark this answers is not in the document]
Comment (qiujian16) [reply pkg/reconciler/apis/apibindingdeletion/apibinding_deletion_controller.go:331]: hrm, I thought we do not need RV/UID precondition when finalization.

[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply pkg/reconciler/apis/apibindingdeletion/apibinding_deletion_controller.go:331]: If we don't, that's fine!

[the remark this answers is not in the document]
Comment (qiujian16) [reply pkg/reconciler/apis/apibindingdeletion/apibinding_deletion_controller.go:333]: Added the uid/rv as precond :)
